extensions/narayana-jta/runtime/src/main/java/io/quarkus/narayana/jta/QuarkusTransactionImpl.java

```diff
@@ -0,0 +1,245 @@
+ */
+@RequestScoped
+public class QuarkusTransactionImpl implements QuarkusTransaction {
+
+    private static final Logger log = Logger.getLogger(QuarkusTransactionImpl.class);
+
+    private final UserTransaction userTransaction;
+    private final TransactionManagerConfiguration transactionManagerConfiguration;
+    boolean active;
+    boolean rollbackOnly;
+    boolean autoCommit;
+
+    @Inject
+    public QuarkusTransactionImpl(UserTransaction userTransaction,
+            TransactionManagerConfiguration transactionManagerConfiguration) {
+        this.userTransaction = userTransaction;
+        this.transactionManagerConfiguration = transactionManagerConfiguration;
+    }
+
+    public QuarkusTransactionImpl() {
+        //for proxiability
+        this.userTransaction = null;
+        this.transactionManagerConfiguration = null;
+    }
+
+    @Override
+    public TransactionBuilder builder() {
+        return new TransactionBuilder() {
+
+            boolean requestScoped;
+            int timeout = 0;
+
+            @Override
+            public void begin() {
+                QuarkusTransactionImpl.this.begin(requestScoped, timeout);
+            }
+
+            @Override
+            public TransactionBuilder requestScoped() {
+                requestScoped = true;
+                return this;
+            }
+
+            @Override
+            public TransactionBuilder timeout(int seconds) {
+                if (timeout < 0) {
+                    throw new IllegalArgumentException("seconds cannot be negative");
+                }
+                timeout = seconds;
+                return this;
+            }
+
+            @Override
+            public void run(Runnable r) {
+                if (requestScoped) {
+                    throw new IllegalStateException("cannot use 'requestScoped' with 'run'");
+                }
+                runInTransaction(timeout, r);
+            }
+
+            @Override
+            public <T> T call(Callable<T> c) {
+                if (requestScoped) {
+                    throw new IllegalStateException("cannot use 'requestScoped' with 'call'");
+                }
+                return callInTransaction(timeout, c);
+            }
+
+        };
+    }
+
+    @Override
+    public void begin() {
+        begin(false, 0);
+    }
+
+    private void begin(boolean autoCommit, int timeout) {
+        try {
+            if (active) {
+                throw new RuntimeException("Transaction already active");
+            }
+            this.autoCommit = autoCommit;
+            if (timeout > 0) {
+                userTransaction.setTransactionTimeout(timeout);
+            }
+            active = true;
+            userTransaction.begin();
+            if (timeout > 0) {
+                userTransaction.setTransactionTimeout(
+                        (int) transactionManagerConfiguration.defaultTransactionTimeout.get().toSeconds());
+            }
```

Comment:
I think the if condition should be ```if (timeout == 0)```. And it could be better to add a timeout test case.

Reply:
I believe Stuart has coded it such that a value of 0 means that the user wants to use the default value configured in the transaction configuration.

Reply:
This is restoring the default after it was set above. Really though it should be in a finally block.

Reply:
Thanks for explanation and it makes sense now.